### PR TITLE
Forced Bucketing

### PR DESCRIPTION
**Pull request by faisal-iqbal**
Forced bucketing feature.

### CODE REVIEW

File: spec/decision_service_spec.rb [hunk 1]
```diff
@@ -109,6 +114,7 @@
     end
 
     it 'should respect forced variations within mutually exclusive grouped experiments' do
+      expect(config.set_forced_variation('group1_exp2', 'forced_group_user1', 'g1_e2_v2')).to eq(TRUE)
```
Comment (msohailhussain) [inline spec/decision_service_spec.rb]: @faisal-iqbal  I will get back to you, it's whitelisted as you mentioned and setforced variation not supposed to be here. It is getting variation from the config file.

File: spec/decision_service_spec.rb [hunk 2]
```diff
@@ -120,11 +126,12 @@
     end
 
     it 'should bucket normally if user is whitelisted into a forced variation that is not in the datafile' do
+      expect(config.set_forced_variation('test_experiment', 'forced_user_with_invalid_variation', 'invalid_variation')).to eq(FALSE)
```
Comment (msohailhussain) [inline spec/decision_service_spec.rb]: same here.

File: spec/decision_service_spec.rb [hunk 3]
```diff
@@ -120,11 +126,12 @@
...
       expect(decision_service.get_variation('test_experiment', 'forced_user_with_invalid_variation')).to eq('111128')
       expect(spy_logger).to have_received(:log)
                             .once.with(
                               Logger::INFO,
-                              "User 'forced_user_with_invalid_variation' is whitelisted into variation 'invalid_variation', which is not in the datafile."
+                              "User 'forced_user_with_invalid_variation' is in variation 'control' of experiment 'test_experiment'."
```
Comment (msohailhussain) [inline spec/decision_service_spec.rb]: please discard it. it's whitelist case not the forced variation case.

File: spec/decision_service_spec.rb (hunk 1 above)
Comment (faisal-iqbal) [reply spec/decision_service_spec.rb]: Removed set_forced_variation from here

File: spec/decision_service_spec.rb (hunk 2 above)
Comment (faisal-iqbal) [reply spec/decision_service_spec.rb]: Removed set_forced_variation from here

File: spec/decision_service_spec.rb (hunk 3 above)
Comment (faisal-iqbal) [reply spec/decision_service_spec.rb]: Discarded this change

File: lib/optimizely.rb [hunk 4]
```diff
@@ -144,6 +144,37 @@ def get_variation(experiment_key, user_id, attributes = nil)
...
+      # Returns String|nil The forced variation key.
+
+      forced_variation_key = nil
+
+      forced_variation = @config.get_forced_variation(experiment_key, user_id);
+      if forced_variation
+        forced_variation_key = forced_variation['key']
+      end
+
+      forced_variation_key
```
Comment (msohailhussain) [inline lib/optimizely.rb]: without return statement, does ruby return? :S

Comment (msohailhussain) [inline lib/optimizely.rb]: no return statement?

File: lib/optimizely/decision_service.rb [hunk 5]
```diff
@@ -289,11 +293,11 @@ def get_forced_variation_id(experiment_key, user_id)
       #
       # Returns variation ID into which user_id is forced (nil if no variation)
 
-      forced_variations = @config.get_forced_variations(experiment_key)
+      forced_variations = @config.get_forced_variation(experiment_key, user_id)
```
Comment (msohailhussain) [inline lib/optimizely/decision_service.rb]: this method should be removed. getforcedvariation from decision service, I assume it should be renamed as getwhitelisted variation which is below.

File: lib/optimizely.rb (hunk 4 above)
Comment (msohailhussain) [reply lib/optimizely.rb]: just knew, it is returned automatically.

**Pull request closed without merging**